src/payload.rs

```diff
@@ -907,8 +908,28 @@ unsafe extern "C" fn z_bytes_writer_write(
     this: &mut z_loaned_bytes_writer_t,
     src: *const u8,
     len: usize,
```

Comment:
maybe it is simpler to take len by mutable reference and just update it ?
Or alternatively return isize and use positive values to indicated the number of bytes not written ?

Reply:
We agreed on this point with Luca and Angelo yesterday.
We cannot return error in any form except z_result_t

Reply:
z_bytes_reader_read returns number of bytes read. It will look very strange if write works differently. 
Also standard fwrite and fread on unix return number of bytes read/written.

Reply:
Reader is infallible while our writer may fail with an error...

Reply:
It is common API for Rust std Writer trait.

Reply:
reader can fail due to the same reason as write (i.e. if user asked for more bytes than available). And these APIs a for c and not for rust.

Reply:
reader will not fail, it will read less than requested (up to 0 bytes)

btw, I guess for writer to be compatible with reader (and common UNIX write-read things) we need to return `len` instead of `result`, and 0 will mean "nothing written" that should be treated as an error

Reply:
After some thinking, that's what I suggest

Add new common type z_io_result_t:
```
/// IO result
/// On io success, contains number of handled bytes 
/// On io error, contains negative z_result_t value
pub type z_io_result_t = isize;
```
Use this type for all read\write io as return type
Pros:
- we get unix-like io API logic
- less arguments compared to adding `bytes_written`
- easy to use

Reply:
Maybe it is easier to just to expose only write_all functionality which so far fits into z_result_t frames ?

Reply:
The only thing I can say here is that i'd like to have full logic in API. Rust has both write and write_all - why not having the same in C?

Reply:
I'm not sure if it makes much difference whether you got a error or not as long as you failed to write all the bytes you wanted, so in my opinion it is much easier just to consider 2 cases: all bytes have been successfully written or not. If the need arises we can always add partial write in the future...

Reply:
Partial write may be OK in many cases, so I'd rather suggest having it already, especially taking into account the fact that we already have it in Rust

Reply:
In case of `z_bytes_t` I agree with @DenisBiryukov91 . In this specific case `z_bytes_t` is used for serialization, so definitely you don't want to send a partially-serialized object. In Rust the write exists because it is part of the standard [Write trait](https://doc.rust-lang.org/std/io/trait.Write.html#) and allows to pass the `ZBytesWriter` to anything that accepts `T: std::io::Write`. 

Long story short, I'm fine at the moment to have only `z_bytes_writer_write_all` and to postpone the discussion for `z_bytes_writer_write` after `1.0.0` release. It would be an addition, so not a breaking change.

Reply:
@Mallets okay, I'll remove write